<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Lauren Sanders, Jian Gong, Vaishnavi Nagesh">

<title>RR9 Retina Dataset Integration</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="rr9_retina_dataset_files/libs/clipboard/clipboard.min.js"></script>
<script src="rr9_retina_dataset_files/libs/quarto-html/quarto.js"></script>
<script src="rr9_retina_dataset_files/libs/quarto-html/popper.min.js"></script>
<script src="rr9_retina_dataset_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="rr9_retina_dataset_files/libs/quarto-html/anchor.min.js"></script>
<link href="rr9_retina_dataset_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="rr9_retina_dataset_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="rr9_retina_dataset_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="rr9_retina_dataset_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="rr9_retina_dataset_files/libs/bootstrap/bootstrap-c0367b04c37547644fece4185067e4a7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-full">

<main class="content column-page" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">RR9 Retina Dataset Integration</h1>
</div>



<div class="quarto-title-meta column-page">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Lauren Sanders, Jian Gong, Vaishnavi Nagesh </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="digital-twin-project-description" class="level2">
<h2 class="anchored" data-anchor-id="digital-twin-project-description">Digital Twin Project Description</h2>
<p>Space biology confronts a critical obstacle: the challenge of incomplete data due to the logistical complexities and high costs of space missions. Addressing this issue, this research presents strategies that integrate AI and digital twin technology to overcome the limitations posed by sparse datasets in space biology research.</p>
<p>By presenting a cohesive strategy that combines synthetic data generation, automatic labeling, and advanced machine learning with digital twins, we showcase an application to the RR9 dataset at OSDR. This research aims to overcome the challenge of data scarcity in space biology, thereby forging a way to unlock insights into the potential of life beyond Earth.</p>
</section>
<section id="rr9-background" class="level2">
<h2 class="anchored" data-anchor-id="rr9-background">RR9 Background</h2>
<p>The Rodent Research 9 payload consisted of three space biology experiments designed to examine impacts of long-duration spaceflight on visual impairment and joint tissue degradation that affect astronauts.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 33%">
<col style="width: 33%">
<col style="width: 33%">
</colgroup>
<thead>
<tr class="header">
<th>Investigation</th>
<th>Purpose</th>
<th>Experiments</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Investigation 1</td>
<td>Effects of microgravity on fluid shifts and increased fluid pressures that occur in the head.</td>
<td>1. To determine whether spaceflight on the ISS alters rodent basilar artery spontaneous tone, myogenic and KCl (Potassium Chloride)-evoked vasoconstriction, mechanical stiffness and gross structure.<br> 2. To estimate whether spaceflight on the ISS alters the blood-brain barrier in rodents, as indicated by ultrastructural examination of the junctional complex of the cerebral capillary endothelium.<br> 3. To determine whether spaceflight on the ISS alters rodent basal vein (inside cranium) and jugular vein (outside cranium) spontaneous tone, myogenic and KCl-evoked constriction, distension, and gross structure.<br> 4. To determine whether spaceflight on the ISS alters the ability of the cervical lymphatics to modulate lymph flow, and thus, regulate cerebral fluid homeostasis.</td>
</tr>
<tr class="even">
<td>Investigation 2</td>
<td>Impact of spaceflight on the vessels that supply blood to the eyes.</td>
<td>1. Define the relationships between spaceflight condition-induced oxidative stress in reactive oxygen species (ROS) expression and retinal vascular remodeling and BRB function in mice return to Earth alive.<br> 2. Determine whether spaceflight condition-induced oxidative damage in retina is mediated through photoreceptor mitochondrial ROS production.</td>
</tr>
<tr class="odd">
<td>Investigation 3</td>
<td>Extent of knee and hip joint degradation caused by prolonged exposure to weightlessness.</td>
<td>1. Determine the extent of knee and hip joint degradation in mice after 30 days of spaceflight on the ISS.<br> 2. Use the DigiGait System to assess gait patterns before and after returning from the ISS.</td>
</tr>
</tbody>
</table>
<p>We are interested in the Retinal data and all things related to Eye. So, all the experiments related to Investigation 1 and 2 will be studied here. Below is the table of all OSD identifiers related to above investigations obtained from https://osdr.nasa.gov/bio/repo/data/payloads/RR-9</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Identifier</th>
<th>Title</th>
<th>Factors</th>
<th>Assay Types</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="https://osdr.nasa.gov/bio/repo/data/studies/OSD-557">OSD-557</a></td>
<td>Spaceflight influences gene expression, photoreceptor integrity, and oxidative stress related damage in the murine retina (RR-9)</td>
<td>Spaceflight</td>
<td>Bone Microstructure<br>Molecular Cellular Imaging<br>histology</td>
</tr>
<tr class="even">
<td><a href="https://osdr.nasa.gov/bio/repo/data/studies/OSD-568">OSD-568</a></td>
<td>Characterization of mouse ocular responses (Microscopy) to a 35-day (RR-9) spaceflight mission: Evidence of blood-retinal barrier disruption and ocular adaptations</td>
<td>Spaceflight</td>
<td>Molecular Cellular Imaging</td>
</tr>
<tr class="odd">
<td><a href="https://osdr.nasa.gov/bio/repo/data/studies/OSD-715">OSD-715</a></td>
<td>Characterization of mouse ocular response to a 35-day spaceflight mission: Evidence of blood-retinal barrier disruption and ocular adaptations - Proteomics data</td>
<td>Spaceflight</td>
<td>protein expression profiling</td>
</tr>
<tr class="even">
<td><a href="https://osdr.nasa.gov/bio/repo/data/studies/OSD-255">OSD-255</a></td>
<td>Spaceflight influences gene expression, photoreceptor integrity, and oxidative stress-related damage in the murine retina</td>
<td>Spaceflight</td>
<td>transcription profiling</td>
</tr>
<tr class="odd">
<td><a href="https://osdr.nasa.gov/bio/repo/data/experiments/OS-140">OS-140</a></td>
<td>Space Flight Environment Induces Remodeling of Vascular Network and Glia-Vascular Communication in Mouse Retina</td>
<td>Spaceflight</td>
<td></td>
</tr>
<tr class="even">
<td><a href="https://osdr.nasa.gov/bio/repo/data/studies/OSD-583">OSD-583</a></td>
<td>Characterization of mouse ocular responses (intraocular pressure) to a 35-day (RR-9) spaceflight mission: Evidence of blood-retinal barrier disruption and ocular adaptations</td>
<td>Spaceflight</td>
<td>Tonometry</td>
</tr>
</tbody>
</table>
<p>The purpose of this notebook is to combine all retina data from the <a href="https://osdr.nasa.gov/bio/repo/data/payloads/RR-9">Rodent Research 9</a> (RR9) mission from the <a href="https://www.nasa.gov/osdr/%20(OSDR)">NASA Open Science Data Repository</a>, perform exploratory data analysis, impute missing data and train a digital twin.</p>
<p><strong>Original Author:</strong> Lauren Sanders</p>
<p><strong>Additional Author(s):</strong> Jian Gong, Vaishnavi Nagesh</p>
</section>
<section id="import-statements" class="level2">
<h2 class="anchored" data-anchor-id="import-statements">Import Statements</h2>
<p>Importing all the required libraries here, declare global variables like source name dictionary.</p>
<div id="e6372eaa" class="cell" data-execution_count="3">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Decalre source name mapping</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Add Source Name to all the dfs that are missing it</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename some Source Names that already exist</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Fviv = CC1 (confirmed in OSD-583 metadata)</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co">#GViv = CC2?</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>GSM_DICT <span class="op">=</span> {</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932702"</span>: <span class="st">"F11"</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932703"</span>: <span class="st">"F15"</span>,</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932704"</span>: <span class="st">"F16"</span>,</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932705"</span>: <span class="st">"F17"</span>,</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932706"</span>: <span class="st">"F18"</span>,</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932707"</span>: <span class="st">"F19"</span>,</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932708"</span>: <span class="st">"F20"</span>,</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932694"</span>: <span class="st">"GC11"</span>,</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932695"</span>: <span class="st">"GC15"</span>,</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932696"</span>: <span class="st">"GC16"</span>,</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932697"</span>: <span class="st">"GC17"</span>,</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932698"</span>: <span class="st">"GC18"</span>,</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932699"</span>: <span class="st">"GC19"</span>,</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932700"</span>: <span class="st">"GC20"</span>,</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932693"</span>: <span class="st">"GC9"</span>,</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>    <span class="st">"GSM3932701"</span>: <span class="st">"F9"</span></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load Data</h2>
<div id="24e37a7c" class="cell" data-cache="true" data-execution_count="4">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Using the Data API from the OSDR Public API</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Documentation: https://www.nasa.gov/reference/osdr-public-api/</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>rnaseq <span class="op">=</span> pd.read_csv(<span class="st">'https://osdr.nasa.gov/geode-py/ws/studies/OSD-255/download?source=datamanager&amp;file=GLDS-255_rna_seq_Normalized_Counts.csv'</span>, index_col<span class="op">=</span><span class="dv">0</span>).transpose()</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>rnaseq[<span class="st">'Source Name'</span>] <span class="op">=</span> rnaseq.index.<span class="bu">map</span>(GSM_DICT)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>zo1 <span class="op">=</span> pd.read_csv(<span class="st">'https://osdr.nasa.gov/geode-py/ws/studies/OSD-568/download?source=datamanager&amp;file=LSDS-5_immunostaining_microscopy_Zo-1tr_TRANSFORMED.csv'</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>zo1[<span class="st">'Source Name'</span>] <span class="op">=</span> [<span class="st">'F15'</span>, <span class="st">'F16'</span>, <span class="st">'F17'</span>, <span class="st">'F18'</span>, <span class="st">'F19'</span>, <span class="st">'F20'</span>, <span class="st">'GC15'</span>, <span class="st">'GC16'</span>, <span class="st">'GC17'</span>, <span class="st">'GC18'</span>, <span class="st">'GC19'</span>, <span class="st">'GC20'</span>]</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>tunel <span class="op">=</span> pd.read_csv(<span class="st">'https://osdr.nasa.gov/geode-py/ws/studies/OSD-568/download?source=datamanager&amp;file=LSDS-5_immunostaining_microscopy_TUNELtr_TRANSFORMED.csv'</span>)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>tunel[<span class="st">'Source Name'</span>] <span class="op">=</span> [<span class="st">'F15'</span>, <span class="st">'F16'</span>, <span class="st">'F17'</span>, <span class="st">'F18'</span>, <span class="st">'F19'</span>, <span class="st">'F20'</span>, <span class="st">'GC15'</span>, <span class="st">'GC16'</span>, <span class="st">'GC17'</span>, <span class="st">'GC18'</span>, <span class="st">'GC19'</span>, <span class="st">'GC20'</span>, <span class="st">'CC2_15'</span>, <span class="st">'CC2_16'</span>, <span class="st">'CC2_17'</span>, <span class="st">'CC2_18'</span>, <span class="st">'CC2_20'</span>, <span class="st">'Viv15'</span>, <span class="st">'Viv16'</span>, <span class="st">'Viv17'</span>, <span class="st">'Viv18'</span>, <span class="st">'Viv19'</span>, <span class="st">'Viv20'</span>] <span class="co">#rename VG to CC2 and V to Viv for consistency</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>pecam <span class="op">=</span> pd.read_csv(<span class="st">'https://osdr.nasa.gov/geode-py/ws/studies/OSD-568/download?source=datamanager&amp;file=LSDS-5_immunostaining_microscopy_PECAMtr_TRANSFORMED.csv'</span>)<span class="co">#['Source Name']=['F15','F16','F17','F18','F19','F20','GC15','GC16','GC17','GC18','GC19']</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>pecam[<span class="st">'Source Name'</span>] <span class="op">=</span> [<span class="st">'F15'</span>, <span class="st">'F16'</span>, <span class="st">'F17'</span>, <span class="st">'F18'</span>, <span class="st">'F19'</span>, <span class="st">'F20'</span>, <span class="st">'GC15'</span>, <span class="st">'GC16'</span>, <span class="st">'GC17'</span>, <span class="st">'GC18'</span>, <span class="st">'GC19'</span>]</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>microct <span class="op">=</span> pd.read_csv(<span class="st">'https://osdr.nasa.gov/geode-py/ws/studies/OSD-557/download?source=datamanager&amp;file=LSDS-1_microCT_MicroCT_Transformed_Reusable_Results.csv'</span>) <span class="co">#NOTE: OSD-557 also has raw image files for immunostaining and H&amp;E</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>microct[<span class="st">'Source Name'</span>] <span class="op">=</span> [<span class="st">'F10'</span>, <span class="st">'F12'</span>, <span class="st">'F13'</span>, <span class="st">'F14'</span>, <span class="st">'Viv10'</span>, <span class="st">'Viv12'</span>, <span class="st">'Viv13'</span>, <span class="st">'Viv14'</span>, <span class="st">'GC10'</span>, <span class="st">'GC11'</span>, <span class="st">'GC13'</span>, <span class="st">'GC14'</span>] <span class="co">#rename V to Viv</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>pna <span class="op">=</span> pd.read_csv(<span class="st">'https://osdr.nasa.gov/geode-py/ws/studies/OSD-557/download?source=datamanager&amp;file=LSDS-1_immunostaining_microscopy_PNAtr_Transformed_Reusable_Results.csv'</span>)</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>hne <span class="op">=</span> pd.read_csv(<span class="st">'https://osdr.nasa.gov/geode-py/ws/studies/OSD-557/download?source=datamanager&amp;file=LSDS-1_immunostaining_microscopy_HNEtr_Transformed_Reusable_Results.csv'</span>)</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>hne[<span class="st">'Source Name'</span>] <span class="op">=</span> [<span class="st">"F15"</span>, <span class="st">"F16"</span>, <span class="st">"F17"</span>, <span class="st">"F18"</span>, <span class="st">"F19"</span>, <span class="st">"F20"</span>, <span class="st">"GC15"</span>, <span class="st">"GC16"</span>, <span class="st">"GC17"</span>, <span class="st">"GC18"</span>, <span class="st">"GC19"</span>, <span class="st">"Viv15"</span>, <span class="st">"Viv16"</span>, <span class="st">"Viv17"</span>, <span class="st">"Viv18"</span>, <span class="st">"Viv19"</span>, <span class="st">"Viv20"</span>, <span class="st">"CC2_15"</span>, <span class="st">"CC2_16"</span>, <span class="st">"CC2_17"</span>, <span class="st">"CC2_18"</span>, <span class="st">"CC2_19"</span>, <span class="st">"CC2_20"</span>]</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>tonometry <span class="op">=</span> pd.read_csv(<span class="st">'https://osdr.nasa.gov/geode-py/ws/studies/OSD-583/download?source=datamanager&amp;file=LSDS-16_tonometry_maoTRANSFORMED.csv'</span>)</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>tonometry[<span class="st">'Source Name'</span>] <span class="op">=</span> [<span class="st">"F1"</span>, <span class="st">"F2"</span>, <span class="st">"F3"</span>, <span class="st">"F4"</span>, <span class="st">"F5"</span>, <span class="st">"F6"</span>, <span class="st">"F7"</span>, <span class="st">"F8"</span>, <span class="st">"F9"</span>, <span class="st">"F10"</span>, <span class="st">"F11"</span>, <span class="st">"F12"</span>, <span class="st">"F13"</span>, <span class="st">"F14"</span>, <span class="st">"F15"</span>, <span class="st">"F16"</span>, <span class="st">"F17"</span>, <span class="st">"F18"</span>, <span class="st">"F19"</span>, <span class="st">"F20"</span>, <span class="st">"CC1_1"</span>, <span class="st">"CC1_2"</span>, <span class="st">"CC1_3"</span>, <span class="st">"CC1_4"</span>, <span class="st">"CC1_5"</span>, <span class="st">"CC1_6"</span>, <span class="st">"CC1_7"</span>, <span class="st">"CC1_8"</span>, <span class="st">"CC1_9"</span>, <span class="st">"CC1_10"</span>, <span class="st">"CC1_11"</span>, <span class="st">"CC1_12"</span>, <span class="st">"CC1_13"</span>, <span class="st">"CC1_14"</span>, <span class="st">"CC1_15"</span>, <span class="st">"CC1_16"</span>, <span class="st">"CC1_17"</span>, <span class="st">"CC1_18"</span>, <span class="st">"CC1_19"</span>, <span class="st">"CC1_20"</span>, <span class="st">"GC1"</span>, <span class="st">"GC2"</span>, <span class="st">"GC3"</span>, <span class="st">"GC4"</span>, <span class="st">"GC5"</span>, <span class="st">"GC6"</span>, <span class="st">"GC7"</span>, <span class="st">"GC8"</span>, <span class="st">"GC9"</span>, <span class="st">"GC10"</span>, <span class="st">"GC11"</span>, <span class="st">"GC12"</span>, <span class="st">"GC13"</span>, <span class="st">"GC14"</span>, <span class="st">"GC15"</span>, <span class="st">"GC16"</span>, <span class="st">"GC17"</span>, <span class="st">"GC18"</span>, <span class="st">"GC19"</span>, <span class="st">"GC20"</span>, <span class="st">"Viv1"</span>, <span class="st">"Viv2"</span>, <span class="st">"Viv3"</span>, <span class="st">"Viv4"</span>, <span class="st">"Viv5"</span>, <span class="st">"Viv6"</span>, <span class="st">"Viv7"</span>, <span class="st">"Viv8"</span>, <span class="st">"Viv9"</span>, <span class="st">"Viv10"</span>, <span class="st">"Viv11"</span>, <span class="st">"Viv12"</span>, <span class="st">"Viv13"</span>, <span class="st">"Viv14"</span>, <span class="st">"Viv15"</span>, <span class="st">"Viv16"</span>, <span class="st">"Viv17"</span>, <span class="st">"Viv18"</span>, <span class="st">"Viv19"</span>, <span class="st">"Viv20"</span>, <span class="st">"CC2_1"</span>, <span class="st">"CC2_2"</span>, <span class="st">"CC2_3"</span>, <span class="st">"CC2_4"</span>, <span class="st">"CC2_5"</span>, <span class="st">"CC2_6"</span>, <span class="st">"CC2_7"</span>, <span class="st">"CC2_8"</span>, <span class="st">"CC2_9"</span>, <span class="st">"CC2_10"</span>, <span class="st">"CC2_11"</span>, <span class="st">"CC2_12"</span>, <span class="st">"CC2_13"</span>, <span class="st">"CC2_14"</span>, <span class="st">"CC2_15"</span>, <span class="st">"CC2_16"</span>, <span class="st">"CC2_17"</span>, <span class="st">"CC2_18"</span>, <span class="st">"CC2_19"</span>, <span class="st">"CC2_20"</span>] <span class="co">#rename "FViv" to "CC1"  for consistency</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>protein <span class="op">=</span> pd.read_excel(<span class="st">'https://osdr.nasa.gov/geode-py/ws/studies/OSD-715/download?source=datamanager&amp;file=GLDS-639_proteomics_01_Mao_RR9_Retina_092018_MQ_iBAQ.xlsx'</span>, sheet_name<span class="op">=</span><span class="st">'Combat corrected'</span>, index_col<span class="op">=</span><span class="dv">0</span>).transpose()</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>protein <span class="op">=</span> protein.drop(protein.columns[[<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>]], axis<span class="op">=</span><span class="dv">1</span>).rename(columns<span class="op">=</span>{<span class="st">"Sample"</span>: <span class="st">"Source Name"</span>}).reset_index().drop(columns<span class="op">=</span>[<span class="st">'index'</span>]) <span class="co"># remove some Excel formatting</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>protein <span class="op">=</span> protein.set_index(<span class="st">'Source Name'</span>).<span class="bu">apply</span>(pd.to_numeric, errors<span class="op">=</span><span class="st">'coerce'</span>) <span class="co"># convert to numeric instead of dtype "object"</span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>protein[<span class="st">'Source Name'</span>] <span class="op">=</span> protein.index</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>protein.reset_index(drop<span class="op">=</span><span class="va">True</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>sc_gp <span class="op">=</span> pd.DataFrame({<span class="st">'Source Name'</span>: [<span class="st">"F1"</span>, <span class="st">"F2"</span>, <span class="st">"F3"</span>, <span class="st">"F4"</span>, <span class="st">"F5"</span>, <span class="st">"F6"</span>, <span class="st">"F7"</span>, <span class="st">"F8"</span>, <span class="st">"F9"</span>, <span class="st">"F10"</span>, <span class="st">"F11"</span>, <span class="st">"F12"</span>, <span class="st">"F13"</span>, <span class="st">"F14"</span>, <span class="st">"F15"</span>, <span class="st">"F16"</span>, <span class="st">"F17"</span>, <span class="st">"F18"</span>, <span class="st">"F19"</span>, <span class="st">"F20"</span>, <span class="st">"CC1_1"</span>, <span class="st">"CC1_2"</span>, <span class="st">"CC1_3"</span>, <span class="st">"CC1_4"</span>, <span class="st">"CC1_5"</span>, <span class="st">"CC1_6"</span>, <span class="st">"CC1_7"</span>, <span class="st">"CC1_8"</span>, <span class="st">"CC1_9"</span>, <span class="st">"CC1_10"</span>, <span class="st">"CC1_11"</span>, <span class="st">"CC1_12"</span>, <span class="st">"CC1_13"</span>, <span class="st">"CC1_14"</span>, <span class="st">"CC1_15"</span>, <span class="st">"CC1_16"</span>, <span class="st">"CC1_17"</span>, <span class="st">"CC1_18"</span>, <span class="st">"CC1_19"</span>, <span class="st">"CC1_20"</span>, <span class="st">"GC1"</span>, <span class="st">"GC2"</span>, <span class="st">"GC3"</span>, <span class="st">"GC4"</span>, <span class="st">"GC5"</span>, <span class="st">"GC6"</span>, <span class="st">"GC7"</span>, <span class="st">"GC8"</span>, <span class="st">"GC9"</span>, <span class="st">"GC10"</span>, <span class="st">"GC11"</span>, <span class="st">"GC12"</span>, <span class="st">"GC13"</span>, <span class="st">"GC14"</span>, <span class="st">"GC15"</span>, <span class="st">"GC16"</span>, <span class="st">"GC17"</span>, <span class="st">"GC18"</span>, <span class="st">"GC19"</span>, <span class="st">"GC20"</span>, <span class="st">"Viv1"</span>, <span class="st">"Viv2"</span>, <span class="st">"Viv3"</span>, <span class="st">"Viv4"</span>, <span class="st">"Viv5"</span>, <span class="st">"Viv6"</span>, <span class="st">"Viv7"</span>, <span class="st">"Viv8"</span>, <span class="st">"Viv9"</span>, <span class="st">"Viv10"</span>, <span class="st">"Viv11"</span>, <span class="st">"Viv12"</span>, <span class="st">"Viv13"</span>, <span class="st">"Viv14"</span>, <span class="st">"Viv15"</span>, <span class="st">"Viv16"</span>, <span class="st">"Viv17"</span>, <span class="st">"Viv18"</span>, <span class="st">"Viv19"</span>, <span class="st">"Viv20"</span>, <span class="st">"CC2_1"</span>, <span class="st">"CC2_2"</span>, <span class="st">"CC2_3"</span>, <span class="st">"CC2_4"</span>, <span class="st">"CC2_5"</span>, <span class="st">"CC2_6"</span>, <span class="st">"CC2_7"</span>, <span class="st">"CC2_8"</span>, <span class="st">"CC2_9"</span>, <span class="st">"CC2_10"</span>, <span class="st">"CC2_11"</span>, <span class="st">"CC2_12"</span>, <span class="st">"CC2_13"</span>, <span class="st">"CC2_14"</span>, <span class="st">"CC2_15"</span>, <span class="st">"CC2_16"</span>, <span class="st">"CC2_17"</span>, <span class="st">"CC2_18"</span>, <span class="st">"CC2_19"</span>, <span class="st">"CC2_20"</span>],</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a><span class="st">'Group'</span>: [<span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"CC1"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"GC"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"Viv"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>, <span class="st">"CC2"</span>]})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="data-exploration-and-validation" class="level2">
<h2 class="anchored" data-anchor-id="data-exploration-and-validation">Data Exploration and Validation</h2>
<div id="4a662044" class="cell" data-cache="true" data-execution_count="5">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Drop "Sample Name" and a couple other irrelevant columns</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>zo1.drop(columns<span class="op">=</span>[<span class="st">'Sample_Name'</span>], inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>tunel.drop(columns<span class="op">=</span>[<span class="st">'Sample_Name'</span>], inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>pecam.drop(columns<span class="op">=</span>[<span class="st">'Sample_Name'</span>], inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>microct.drop(columns<span class="op">=</span>[<span class="st">'Sample Name'</span>, <span class="st">'Treatment'</span>], inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>pna.drop(columns<span class="op">=</span>[<span class="st">'Sample Name'</span>, <span class="st">'Treatment'</span>], inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>hne.drop(columns<span class="op">=</span>[<span class="st">'Sample Name'</span>], inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>tonometry.drop(columns<span class="op">=</span>[<span class="st">'Sample Name'</span>, <span class="st">'Factor Value: Spaceflight'</span>, <span class="st">'time_Start'</span>, <span class="st">'Time_End'</span>], inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Add suffix to all physiological data column names to avoid duplicates (rnaseq and protein should be unique)</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>zo1.columns <span class="op">=</span> [col <span class="op">+</span> <span class="st">'_zo1'</span> <span class="cf">if</span> col <span class="op">!=</span> <span class="st">'Source Name'</span> <span class="cf">else</span> col <span class="cf">for</span> col <span class="kw">in</span> zo1.columns]</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>tunel.columns <span class="op">=</span> [col <span class="op">+</span> <span class="st">'_tunel'</span> <span class="cf">if</span> col <span class="op">!=</span> <span class="st">'Source Name'</span> <span class="cf">else</span> col <span class="cf">for</span> col <span class="kw">in</span> tunel.columns]</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>pecam.columns <span class="op">=</span> [col <span class="op">+</span> <span class="st">'_pecam'</span> <span class="cf">if</span> col <span class="op">!=</span> <span class="st">'Source Name'</span> <span class="cf">else</span> col <span class="cf">for</span> col <span class="kw">in</span> pecam.columns]</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>microct.columns <span class="op">=</span> [col <span class="op">+</span> <span class="st">'_microct'</span> <span class="cf">if</span> col <span class="op">!=</span> <span class="st">'Source Name'</span> <span class="cf">else</span> col <span class="cf">for</span> col <span class="kw">in</span> microct.columns]</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>pna.columns <span class="op">=</span> [col <span class="op">+</span> <span class="st">'_pna'</span> <span class="cf">if</span> col <span class="op">!=</span> <span class="st">'Source Name'</span> <span class="cf">else</span> col <span class="cf">for</span> col <span class="kw">in</span> pna.columns]</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>hne.columns <span class="op">=</span> [col <span class="op">+</span> <span class="st">'_hne'</span> <span class="cf">if</span> col <span class="op">!=</span> <span class="st">'Source Name'</span> <span class="cf">else</span> col <span class="cf">for</span> col <span class="kw">in</span> hne.columns]</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>tonometry.columns <span class="op">=</span> [col <span class="op">+</span> <span class="st">'_tonometry'</span> <span class="cf">if</span> col <span class="op">!=</span> <span class="st">'Source Name'</span> <span class="cf">else</span> col <span class="cf">for</span> col <span class="kw">in</span> tonometry.columns]</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>dfs <span class="op">=</span> [tonometry,rnaseq, protein, zo1, tunel, pecam, microct, pna, hne, sc_gp]</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert 'Source Name' to string in all DataFrames to avoid dtype conflicts</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> df <span class="kw">in</span> dfs:</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'Source Name'</span>] <span class="op">=</span> df[<span class="st">'Source Name'</span>].astype(<span class="bu">str</span>)</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>    df.set_index(<span class="st">'Source Name'</span>,inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge all DataFrames on "Source Name" column</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>rr9_all_df <span class="op">=</span> pd.concat(dfs, axis<span class="op">=</span><span class="dv">1</span>, join<span class="op">=</span><span class="st">'outer'</span>)</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>rr9_all_df[<span class="st">'Source Name'</span>] <span class="op">=</span> rr9_all_df.index</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>rr9_all_df <span class="op">=</span> rr9_all_df.reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>rr9_all_df[<span class="st">'Group'</span>] <span class="op">=</span> rr9_all_df[<span class="st">'Group'</span>].astype(<span class="st">'category'</span>)</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Add groups feature to the df.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="summary-of-the-merged-data-frame" class="level2">
<h2 class="anchored" data-anchor-id="summary-of-the-merged-data-frame">Summary of the Merged Data frame</h2>
<div id="2ed00355" class="cell" data-cache="true" data-execution_count="6">
<div class="cell-output cell-output-stderr">
<pre><code>/var/folders/64/f2__yw2s6dxc5bxsxc_9_0xc0000gn/T/ipykernel_32080/2234733613.py:21: FutureWarning:

DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="rr9_retina_dataset_files/figure-html/cell-7-output-2.png" width="1434" height="551" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>